### ISSUE TITLE
ALSA: usb-audio: erroneous downstream patch on dBmin/dBmax calculation

### ISSUE THREAD

**Issue by tomty89** [sonyxperiadev/kernel#1106]
So I have been using a USB DAC (FiiO K1) with my Xperia X and when I check the proc file `usbmixer`, I see that the driver-calculated `dBmin` is doubled to -110.00dB (-11000), where it should be -55.00dB as can be seen under desktop Linux or Windows.

```
u0_a210@F5122:/ $ cat /proc/asound/card1/usbmixer
USB Mixer: usb_id=0x262a100c, ctrlif=1, ctlerr=0
Card: FiiO FiiO USB DAC K1 at usb-msm_hsusb_host-1, full speed
  Unit: 9
    Control: name="PCM Playback Volume", index=0
    Info: id=9, control=2, cmask=0x3, channels=2, type="S16"
    Volume: min=-14080, max=0, dBmin=-11000, dBmax=0
  Unit: 9
    Control: name="PCM Playback Switch", index=0
    Info: id=9, control=1, cmask=0x0, channels=1, type="INV_BOOLEAN"
    Volume: min=0, max=1, dBmin=0, dBmax=0
u0_a210@F5122:/ $
```

Turns out that the Sony kernel is incorporated with an apparently-downstream patch (c234ff1) which erroneously uses the volume resolution (128, i.e. 0.5dB in my case) to do the conversion from the device-reported values (in the scale that is specified in the UAC spec: 1 = 1/256dB) to ALSA value (in the scale of 1 = 1/100dB).  (Hence the issue cannot be noticed with devices with volume resolution of 256, i.e. 1dB.)

The volume resolution (a.k.a. volume stepping) should have nothing to do with the calculation, since it itself should be converted accordingly as well. For example in my case it would be `128 / 256  * 100 = 50`. I am not aware that the ALSA usb audio driver has its volume stepping fixed to 100 (1dB) either. Neither does it have a reason to anyway.

In other words, neither of these should NOT be the case:
dBmin = -11000 / dBmax = 0 / volume stepping in ALSA = 100 (hardcoded) / 110 steps
dBmin = -11000 / dBmax = 0 / volume stepping in ALSA = 50 (calculated from `cval->res`)  / 220 steps

Instead this should be:
dBmin = -5500 / dBmax = 0 / volume stepping in ALSA = 50 (calculated from `cval->res`)  / 110 steps

**Comment by tomty89** [comment]
Fixed with https://github.com/sonyxperiadev/kernel/pull/1107.
